test/unit/canvas_dispose.js

```diff
@@ -0,0 +1,252 @@
+        const lowerCanvas = canvas.lowerCanvasEl;
+        assert.equal(lowerCanvas.getAttribute('data-fabric'), 'main', 'lowerCanvasEl should be marked by fabric');
+        await canvas.dispose();
+        assert.equal(canvas.destroyed, true, 'dispose should flag destroyed');
+        assert.equal(canvas.getObjects().length, 0, 'dispose should clear canvas');
+        assert.equal(canvas.lowerCanvasEl, null, 'dispose should clear lowerCanvasEl');
+        assert.equal(lowerCanvas.hasAttribute('data-fabric'), false, 'dispose should clear lowerCanvasEl data-fabric attr');
+        assert.equal(canvas.contextContainer, null, 'dispose should clear contextContainer');
+    });
+
+    QUnit.test('dispose edge case: multiple calls', async function (assert) {
+        const done = assert.async();
+        const canvas = new klass(null, { renderOnAddRemove: false });
+        assert.notOk(canvas.destroyed, 'should not have been destroyed yet');
+        const res = await Promise.all([
+            canvas.dispose(),
+            canvas.dispose(),
+            canvas.dispose(),
+        ]);
+        assert.ok(canvas.disposed, 'should have flagged `disposed`');
+        assert.ok(canvas.destroyed, 'should have flagged `destroyed`');
+        assert.deepEqual(res, [true, false, false], 'should have disposed in the first call');
+        done();
+    });
+
+    QUnit.test('dispose edge case: multiple calls after `requestRenderAll', async function (assert) {
+        const done = assert.async();
+        const canvas = new klass(null, { renderOnAddRemove: false });
+        assert.notOk(canvas.destroyed, 'should not have been destroyed yet');
+        canvas.requestRenderAll();
+        const res = await Promise.allSettled([
+            canvas.dispose(),
+            canvas.dispose(),
+            canvas.dispose(),
+        ]);
+        assert.ok(canvas.disposed, 'should have flagged `disposed`');
+        assert.ok(canvas.destroyed, 'should have flagged `destroyed`');
+        assert.deepEqual(res, [
+            { status: 'rejected', reason: 'aborted' },
+            { status: 'rejected', reason: 'aborted' },
+            { status: 'fulfilled', value: true }
+        ], 'should have disposed in the last call, aborting the other calls');
+        done();
+    });
+
+    QUnit.test('dispose edge case: rendering after dispose', async function (assert) {
+        const done = assert.async();
+        const canvas = new klass(null, { renderOnAddRemove: false });
+        let called = 0;
+        assert.ok(await canvas.dispose(), 'should dispose');
+        canvas.on('after:render', () => {
+            called++;
+        })
+        canvas.fire('after:render');
+        assert.equal(canvas.nextRenderHandle, undefined);
+        canvas.requestRenderAll();
+        assert.equal(canvas.nextRenderHandle, undefined, '`requestRenderAll` should have no affect');
+        assert.equal(called, 1, 'should not have rendered');
```

Comment:
this assertion here is not really valid since requestRenderAll is async, the called could have been incremented before or after.

Reply:
118c8613